core/src/main/java/io/github/fourlastor/game/demo/state/machine/AttackMelee.java

```diff
@@ -85,21 +96,60 @@ public Action setupAttackAnimation(float distance, float rotationDegrees) {
         return AttackAnimation.makeSequence(source.actor, runnables, positions, moveDuration, rotationDegrees, scale);
     }
 
-    @Override
-    public void enter(GameState state) {
-        Vector2 originalPosition = new Vector2(source.getActorPosition());
-        target.changeHp(-damage);
+    public void doAttackAnimation(Vector2 originalPosition, Vector2 targetPosition) {
         // Distance between source and target is used to scale the animation if needed.
-        float distance = source.getActorPosition().dst(target.getActorPosition());
+        float distance = source.getActorPosition().dst(targetPosition);
         // Angle offset of target from source.
-        float rotationDegrees = calculateAngle(source.getActorPosition(), target.getActorPosition());
+        float rotationDegrees = calculateAngle(originalPosition, targetPosition);
         SequenceAction attackAnimation = Actions.sequence(
                 setupAttackAnimation(distance, rotationDegrees),
                 Actions.run(() -> source.setActorPosition(originalPosition)),
                 Actions.run(router::pickMonster));
         source.actor.addAction(attackAnimation);
     }
 
+    public void attackUnit() {
+        targetUnit.changeHp(-damage);
+        Vector2 originalPosition = new Vector2(source.getActorPosition());
+        Vector2 targetPosition = targetUnit.getActorPosition();
+        doAttackAnimation(originalPosition, targetPosition);
+    }
+
+    public void smashTile() {
+        targetTile.type = TileType.TERRAIN;
+
+        // Actor-related.
+        Actor oldActor = targetTile.actor;
+        Stage stage = oldActor.getStage();
+        YSort ySort = null;
+        for (Actor actor : stage.getActors()) {
+            if (actor instanceof YSort) {
+                ySort = (YSort) actor;
+                break;
+            }
+        }
+        if (ySort != null) {
+            targetTile.actor = new TileOnMap(textureAtlas.findRegion("tiles/white"));
+            targetTile.actor.setPosition(oldActor.getX(), oldActor.getY());
+            ySort.addActor(targetTile.actor);
+            oldActor.remove();
+            ySort.sortChildren();
+        }
```

Comment:
I think there's a bug in here - the tile should become non-solid

you don't need to remove the old/add a new actor - you can set a change the image on the previous one with `actor.setDrawable(new TextureRegionDrawable(textureRegion))`

Reply:
I will try that. How do you want YSort to be interacted with?

Reply:
I think if you don't remove/add it you are already good without touching it

Reply:
oic